레이블 화면 구현 완료

nalaworld opened the pull request:
#### Issue Number
Close #23 
Close #24 
Close #38 

### 변경사항
레이블 목록 화면과 레이블 수정/추가 화면 연결
레이블 수정/추가 화면은 하나의 xib파일로 재사용이 가능하도록 구현
추가화면의 default color 검은색으로 수정

### 새로운 기능
새 레이블 추가
기존의 레이블 정보 수정
레이블 추가/수정 시 정보가 충분히 입력되지 않았을 경우 Alert를 띄워 사용성을 높임
레이블 정보 초기화 기능 구현

|레이블 추가|레이블 정보 수정|레이블 수정 시 textField를 이용|
|------|---|---|
|![addlabel+alert](https://user-images.githubusercontent.com/35067611/97428565-7fb0a980-1959-11eb-9044-2b86a51cb25e.gif)|![editlabel](https://user-images.githubusercontent.com/35067611/97428581-863f2100-1959-11eb-9b5a-e479fe32c610.gif)|![editcolor-keyboard](https://user-images.githubusercontent.com/35067611/97428577-85a68a80-1959-11eb-88a8-dc14df599309.gif)|

### TODO
테스트케이스 생성 후 테스트 진행

### 작업 유형
- [x] 신규 기능 추가
- [x] 버그 수정
- [ ] 리펙토링
- [ ] 문서 업데이트

### 체크리스트
- [x] Merge 하는 브랜치가 올바른가?
- [x] 코딩컨벤션을 준수하는가?
- [x] PR과 관련없는 변경사항이 없는가?
- [x] 내 코드에 대한 자기 검토가 되었는가?
- [ ] 변경사항이 효과적이거나 동작이 작동한다는 것을 보증하는 테스트를 추가하였는가?
- [ ] 새로운 테스트와 기존의 테스트가 변경사항에 대해 만족하는가?

nalaworld committed:
fix :🔥 레이블 cell title 크기 버그 수정

nalaworld committed:
feat :⚡️ 레이블 추가화면 - 레이블 목록화면 연결

nalaworld committed:
feat :⚡️ 레이블 추가 기능 구현 완료

- resolve #23

nalaworld committed:
feat :⚡️ 레이블 수정 기능 구현

- 레이블 수정-추가를 위한 재사용 가능한 xib 파일 생성
- 색상 UILabel -> textField로 변경
- 저장 버튼 클릭 시, 수정 및 추가 분기에 따라 수행

nalaworld committed:
feat :⚡️ 레이블 생성/추가 시 정보를 모두 입력하도록 유도

- 제목, 설명, 색상 중 하나라도 정보가 입력되지 않았다면 alert를 띄우도록 구현

nalaworld committed:
feat :⚡️ 레이블 추가/수정 시 색상 랜덤 생성 구현

SHIVVVPP commented on iOS/IssueTracker/IssueTracker/Label/View/Form/LabelSubmitFormView.swift:
```diff
@@ -9,29 +9,65 @@
+        } else {
+            let alert = UIAlertController(title: "제목, 설명, 색상을\n모두 입력해주세요!", message: nil, preferredStyle: .alert)
+            alert.addAction(UIAlertAction(title: NSLocalizedString("OK", comment: "Default action"), style: .default, handler: nil))
+            self.window?.rootViewController?.present(alert, animated: true, completion: nil)
```
이후에 계층구조를 타고 올라가서 하는 방법 외에 다른 방법을 찾아봐야 할 것 같아요 😢
전에 스프린트할 때 계층구조를 타고 올라가는 방법은 좋지 않다고 리뷰 받은적이 있어서...

nalaworld replied:
계층구조를 타고 올라간다는게 어떤 경우인가요??

SHIVVVPP replied:
`view.superView.superView~` 이런식으로 위로 타고 올라가는 느낌인 것 같아요!

> view -> window -> app delegate을 거쳐 필요한 객체를 가져오는 것은 좋지 않습니다. 뷰의 하이라키와 데이터 모델의 로직이 섞이게 되어 망가지기 쉬운 로직입니다.
이 뷰컨이 vendingMachine을 어디서/어떻게 가져와야할지 다시 정하는게 좋겠습니다.
일단 제가 제안하는 방법은 앱이 시작할때 AppDelegate에서 VendingMachineViewController에 vendingMachine 객체를 할당해주는 것입니다.
이 뷰컨은 vendingMachine 객체가 꼭 있어야 동작하는데 외부에서 옵셔널로 가져와서 동작해야 한다면 모든 곳에 이런 guard를 넣어야하고 무엇보다 뷰 하이라키를 거쳐야 데이터 모델을 가져올 수 있다는 점에서 역할 분리가 제대로 안되어있는 것입니다.


자판기 미션할 때 받았던 리뷰인데... 
지금 보니까 계층구조를 올라가서 데이터 모델에 접근하지 말라는 뜻이었는데 잘못 기억하고 있었나봐요 😄 ㅜㅜ

nalaworld replied:
아하! 저 리뷰는 약간 dependency injection을 사용하는게 좋다는 뉘앙스 같긴하네요! ㅎㅎ 
흠 저희 코드 상에서는 따로 데이터 모델을 건드리는 로직은 없고 딱 alert 창을 띄워주기 위한 window 접근인데 어떨까요?

SHIVVVPP replied:
이슈로 기록해두고 시간 가지고 생각해 보는게 좋을 것 같아요! 🤔

sihyungyou replied:
좋습니다 ㅎㅎ

SHIVVVPP commented on iOS/IssueTracker/IssueTracker/Label/ViewModel/LabelsViewModel.swift:
```diff
@@ -13,13 +13,28 @@ protocol LabelsViewModelProtocol {
     func cellForItemAt(path: IndexPath) -> LabelItemViewModel
     func numberOfItem() -> Int
+    func didAddNewLabel(title: String, desc: String, hexColor: String)
+    func didEditLabel(at indexPath: IndexPath, title: String, desc: String, hexColor: String)
```
생각해보니까 `didAddNewLabel`이랑 `didEditLabel`은 이름이 좀 어색하게 느껴지네요? 🤔 
나중에 서버랑 연결을 생각해 볼 때
상황에 따라서 Label이 추가 안될 수 도 있는데 `did`는 너무 단정적인 느낌이에요!

nalaworld replied:
혹시 어떤 경우에 추가나 수정이 안될까요? 
수정은 indexPath가 확실히 있는 경우로 처리를 해주어서 없는 메모리에 접근하는 시나리오는 예방할 수 있을 것 같고, didAdd, didEdit 함수는 단순히 배열에 추가하거나 배열 속 element의 값을 바꾸는 거라서 이 두 함수 내에서는 다른 에러 시나리오가 없을 것 같아요!

SHIVVVPP commented on iOS/IssueTracker/IssueTracker/Label/View/LabelHeader/LabelHeaderView.swift:
```diff
@@ -10,10 +10,15 @@ import UIKit
 
 class LabelHeaderView: UICollectionReusableView {
     @IBOutlet weak var headerLabel: UILabel!
+    var plusButtonTapped: (() -> Void)?
```
여기 `weak` 설정해줘야 할 것 같네요 ❗

SHIVVVPP commented on iOS/IssueTracker/IssueTracker/Label/View/Form/LabelSubmitFormView.swift:
```diff
@@ -9,29 +9,65 @@
 class LabelSubmitFormView: UIView {
-
+    
+    var submitbuttonTapped: ((String, String, String) -> Void)?
```
`weak` 설정해줘야 될 것 같아요!!

nalaworld replied:
아 클로저 변수들 weak 설정해서 다시 커밋해야겠네요 ㅎㅎ

SHIVVVPP replied:
저희가 좀 급하게 했었나 보네요  😂

nalaworld commented on iOS/IssueTracker/IssueTracker/Label/View/Form/LabelSubmitFormView.swift:
```diff
@@ -9,29 +9,65 @@
+        //        colorView.layer.backgroundColor = defaultColorCode.color
     }
     
     @IBAction func submitButtonTapped(_ sender: UIButton) {
```
아 그리구 submitButtonTapped -> saveButtonTapped 는 어떨까요? 코드를 다시한번 쭉 보고있는데 좀 더 직관적인것 같아서요!

SHIVVVPP replied:
오 좋은 것 같아요! 그런데
추가랑 수정 이 둘을 같이 표현하는 단어를 찾다가 submit을 사용했는데
Form을 제출한다 라는 표현이 더 괜찮지 않을까요?

nalaworld commented on iOS/IssueTracker/IssueTracker/Label/ViewModel/LabelsViewModel.swift:
```diff
@@ -13,13 +13,28 @@ protocol LabelsViewModelProtocol {
+    func didEditLabel(at indexPath: IndexPath, title: String, desc: String, hexColor: String)
 }
 
 class LabelsViewModel: LabelsViewModelProtocol {
```
VC같은 다른 파일들에서 이름이 LableList ~~ 라서 여기를 Labels -> LableList로 하는 게 덜 헷갈릴 것 같네요 다시보니 ㅎㅎ

SHIVVVPP replied:
네 LabelList로 변경하는게 더 좋을 것 같네요!
참! 그리고 group 이름도 Label 에서 Labellist로 바꾸는게 좋을 것 같아요

nalaworld committed:
style :👠 코드리뷰 반영 - 네이밍 수정

ViewModel
- LabelsViewModel -> LabelListViewModel

Method
- didAddNewLabel -> addNewLabel
- didEditLabel -> editLabel

Group
- Label -> LabelScene
